File: Stacks and Queues/CircularQueue.c
#include <stdio.h>
#include <stdlib.h>

#define SIZE 5  // Define the maximum size of the queue

// Circular Queue structure
struct CircularQueue {
    int items[SIZE];
    int front;
    int rear;
};

// Initialize the queue by setting front and rear to -1
void initializeQueue(struct CircularQueue* cq) {
    cq->front = -1;
    cq->rear = -1;
}

// Check if the queue is full
int isFull(struct CircularQueue* cq) {
    return (cq->front == (cq->rear + 1) % SIZE);
}

// Check if the queue is empty
int isEmpty(struct CircularQueue* cq) {
    return (cq->front == -1);
}

// Enqueue operation to insert an element into the queue
void enqueue(struct CircularQueue* cq, int value) {
    if (isFull(cq)) {
        printf("Queue is full!\n");
    } else {
        // If inserting the first element, set front = 0
        if (cq->front == -1)
            cq->front = 0;
        
        // Update rear to the next position in a circular way
        cq->rear = (cq->rear + 1) % SIZE;
        cq->items[cq->rear] = value;
        printf("Inserted %d\n", value);
    }
}

// Dequeue operation to remove an element from the queue
int dequeue(struct CircularQueue* cq) {
    int element;
    if (isEmpty(cq)) {
        printf("Queue is empty!\n");
        return -1;
    } else {
        element = cq->items[cq->front];  // Get the front element
        
        // If the queue has only one element, reset to empty
        if (cq->front == cq->rear) {
            cq->front = -1;
            cq->rear = -1;
        } else {
            // Move front to the next position in a circular way
            cq->front = (cq->front + 1) % SIZE;
        }
        printf("Deleted %d\n", element);
        return element;
    }
}

// Function to display the queue
void display(struct CircularQueue* cq) {
    if (isEmpty(cq)) {
        printf("Queue is empty!\n");
    } else {
        int i = cq->front;
        printf("Queue elements: ");
        while (i != cq->rear) {
            printf("%d ", cq->items[i]);
            i = (i + 1) % SIZE;
        }
        printf("%d\n", cq->items[i]);  // Print the last element (rear)
    }
}

int main() {
    struct CircularQueue cq;
    initializeQueue(&cq);

    enqueue(&cq, 10);
    enqueue(&cq, 20);
    enqueue(&cq, 30);
    enqueue(&cq, 40);
    enqueue(&cq, 50);  // This will fill the queue
    
    display(&cq);  // Display the queue
    
    dequeue(&cq);  // Remove element
    dequeue(&cq);  // Remove element
    
    display(&cq);  // Display the queue after deletion
    
    enqueue(&cq, 60);  // Add new element
    display(&cq);  // Display the updated queue

    return 0;
}

File: Stacks and Queues/CircularQueue.py
class CircularQueue:
    def __init__(self, size):
        """Initialize the queue with a fixed size"""
        self.size = size
        self.queue = [None] * size  # Initialize the queue with None
        self.front = -1  # Front pointer
        self.rear = -1   # Rear pointer

    def is_full(self):
        """Check if the queue is full"""
        return (self.front == (self.rear + 1) % self.size)

    def is_empty(self):
        """Check if the queue is empty"""
        return self.front == -1

    def enqueue(self, value):
        """Insert an element into the circular queue"""
        if self.is_full():
            print("Queue is full!")
        else:
            # If inserting the first element, set front to 0
            if self.front == -1:
                self.front = 0
            
            # Move rear to the next position in a circular way
            self.rear = (self.rear + 1) % self.size
            self.queue[self.rear] = value
            print(f"Inserted {value}")

    def dequeue(self):
        """Remove and return an element from the circular queue"""
        if self.is_empty():
            print("Queue is empty!")
            return None
        else:
            value = self.queue[self.front]  # Get the front element
            
            # If there's only one element, reset the queue
            if self.front == self.rear:
                self.front = -1
                self.rear = -1
            else:
                # Move front to the next position in a circular way
                self.front = (self.front + 1) % self.size
            print(f"Deleted {value}")
            return value

    def display(self):
        """Display the elements of the queue"""
        if self.is_empty():
            print("Queue is empty!")
        else:
            print("Queue elements:", end=" ")
            i = self.front
            while i != self.rear:
                print(self.queue[i], end=" ")
                i = (i + 1) % self.size
            print(self.queue[i])  # Print the last element (rear)


# Test the CircularQueue class
if __name__ == "__main__":
    cq = CircularQueue(5)  # Initialize queue of size 5

    cq.enqueue(10)
    cq.enqueue(20)
    cq.enqueue(30)
    cq.enqueue(40)
    cq.enqueue(50)  # This will fill the queue

    cq.display()  # Display the queue
    
    cq.dequeue()  # Remove an element
    cq.dequeue()  # Remove another element
    
    cq.display()  # Display the queue after deletion
    
    cq.enqueue(60)  # Insert new element
    cq.display()  # Display the updated queue
